src/v/cloud_topics/level_one/frontend_reader/reader.cc

```diff
@@ -0,0 +1,415 @@
+        throw std::runtime_error(
+          fmt::format(
+            "Invalid reader state (ready/empty) for {} ({})", _ntp, _tidp));
+    case state::materialized:
+        consume_materialized_batches(&res);
+        [[fallthrough]];
+    case state::end_of_stream:
+        break;
+    }
+    co_return res;
+}
+
+ss::future<> level_one_log_reader_impl::fetch_metadata(
+  model::timeout_clock::time_point /*deadline*/) {
+    vassert(
+      _state == state::empty,
+      "Invalid state for metadata fetch: {}",
+      std::to_underlying(_state));
+
+    vassert(
+      !_current_obj.has_value(),
+      "Empty state should not have a current object");
+
+    if (_config.start_offset > _config.max_offset) {
+        vlog(
+          cd_log.debug,
+          "L1 reader start_offset {} > max_offset {} for {} ({}): ending "
+          "stream",
+          _config.start_offset,
+          _config.max_offset,
+          _ntp,
+          _tidp);
+        _state = state::end_of_stream;
+        co_return;
+    }
+
+    if (_config.over_budget) {
+        vlog(
+          cd_log.debug,
+          "L1 reader over byte budget for {} ({}): ending stream",
+          _ntp,
+          _tidp);
+        _state = state::end_of_stream;
+        co_return;
+    }
+
+    auto response_fut = co_await ss::coroutine::as_future(
+      _metastore->get_first_ge(_tidp, _config.start_offset));
+    if (response_fut.failed()) {
+        vlog(
+          cd_log.error,
+          "Exception fetching metadata from metastore for {} ({}): {}",
+          _ntp,
+          _tidp,
+          response_fut.get_exception());
+        _state = state::end_of_stream;
+        co_return;
+    }
+
+    auto response = response_fut.get();
+    if (!response.has_value()) {
+        if (response.error() == l1::metastore::errc::out_of_range) {
+            vlog(
+              cd_log.debug,
+              "No L1 objects found for {} ({}) at offset {} or later",
+              _ntp,
+              _tidp,
+              _config.start_offset);
+            _state = state::end_of_stream;
+            co_return;
+        }
+        vlog(
+          cd_log.warn,
+          "Failed to query metastore for {} ({}) offset {}: {}",
+          _ntp,
+          _tidp,
+          _config.start_offset,
+          response.error());
+        _state = state::end_of_stream;
+        co_return;
+    }
+
+    auto& obj = response.value();
+    vlog(
+      cd_log.debug,
+      "Found L1 object {} for {} ({}) at offset {}",
+      obj.oid,
+      _ntp,
+      _tidp,
+      _config.start_offset);
+
+    auto footer_result = co_await read_footer(
+      obj.oid, obj.footer_pos, obj.object_size);
+    if (!footer_result.has_value()) {
+        vlog(
+          cd_log.warn,
+          "Failed to read footer from L1 object {} for {} ({}): {}",
+          obj.oid,
+          _ntp,
+          _tidp,
+          std::to_underlying(footer_result.error()));
+        _state = state::end_of_stream;
+        co_return;
```

Comment:
I do not think we want to silently truncate the stream when there are errors, but instead throw or return some error so the client gets an error